### PR TITLE
MB-61985: Support triggering Scorch events externally

### PR DESCRIPTION
**Pull request by CascadingRadium**
- Add an optional method for Index implementations to support, which would allow firing scorch events externally.

### CODE REVIEW

File: index.go [hunk 1]
```diff
@@ -56,6 +56,13 @@ type CopyIndex interface {
 	// to handle necessary bookkeeping, instead of using the regular IndexReader.
 	CopyReader() CopyReader
 }
+type EventIndex interface {
+	Index
+
+	OnIndexStart()
```
Comment (abhinavdangeti) [inline index.go:62]: Add commentary for each API within the interface.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]
Also, I feel both APIs can do with better naming the current ones are a bit distasteful - I feel there's no need to use the word "Index" in the API for interface that already has it in it's name.

File: index.go (hunk 1 above)
Comment (CascadingRadium) [reply index.go]: refactored now and added comments

File: index.go [hunk 2]
```diff
@@ -57,6 +57,20 @@ type CopyIndex interface {
 	CopyReader() CopyReader
 }
 
+// EventKind represents an event code that can be used to execute external callbacks
+// for various events in the index.
+type EventKind int
+
+// EventKindIndex is a type of event that is fired when a new document is added to the index.
+var EventKindIndex = EventKind(1)
```
Comment (abhinavdangeti) [inline index.go:65]: This is not good @CascadingRadium - 'cos now we have events declared in two places. Let's avoid this kind of detail bleed across libraries. Let's move this back into `bleve`.

File: index.go [hunk 3]
```diff
@@ -56,12 +56,19 @@ type CopyIndex interface {
...
 
-	OnIndex()
+// EventKindIndex is a type of event that is fired when a new document is added to the index.
+var EventKindIndex = EventKind(1)
+
+// EventIndex is an optional interface for exposing the support for firing event
+// callbacks for various events in the index.
+type EventIndex interface {
+	// FireEvent is used to fire an event callback for the specified event kind.
+	FireEvent(event EventKind)
```
Comment (abhinavdangeti) [inline index.go:71]: Can the attribute here be made an `int` instead?

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

File: index.go (hunk 2 above)
Comment (abhinavdangeti) [reply index.go:65]: Also looking into the bleve change, I don't quite see the need for this at all. So maybe you can simply drop this for now, and the `FireEvent` API below will not take any attributes.

**Comment by abhinavdangeti** [comment]
@CascadingRadium Can you look into the commentary here as well. I'll create a tag here once this commit is merged as well.

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**